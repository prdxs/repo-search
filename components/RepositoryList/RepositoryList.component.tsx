import React from 'react';
import clsx from 'clsx';
import styled from 'styled-components';
import Grid from '@material-ui/core/Grid';

import { IRepositoryListProps } from './RepositoryList.typings';
import RepositoryCard from '@/components/RepositoryCard';

const RepositoryList: React.FC<IRepositoryListProps> = ({
  className,
  style,
  repositories,
}) => (
  <Grid
    className={clsx('RepositoryList', className)}
    style={style}
    container
    spacing={4}
  >
    {repositories.map((repo) => (
      <Grid item key={repo.id}>
        <RepositoryCard
          name={repo.name}
          link={repo.link}
          stars={repo.stars}
          forks={repo.forks}
          issues={repo.issues}
          lastUpdated={repo.lastUpdated}
        />
      </Grid>
    ))}
  </Grid>
);

const StyledRepositoryList = styled(RepositoryList)`
  &.RepositoryList {
    max-width: 500px;
    padding: ${({ theme }) => theme.spacing(0, 2, 8)};
  }
`;

export default StyledRepositoryList;
