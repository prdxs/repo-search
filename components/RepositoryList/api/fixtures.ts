import { IRepository } from './typings';

export const repositories: Array<IRepository> = [
  {
    id: '0',
    name: 'facebook/react',
    link: 'https://github.com/facebook/react',
    stars: 151026,
    forks: 29654,
    issues: 498,
    lastUpdated: '1st Jul 2020 07:20:59',
  },
  {
    id: '1',
    name: 'prdxs/repo-search',
    link: 'https://github.com/prdxs/repo-search',
    stars: 0,
    forks: 0,
    issues: 0,
    lastUpdated: '1st Jul 2020 14:20:17',
  },
];
