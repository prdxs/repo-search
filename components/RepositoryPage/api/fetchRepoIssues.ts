import { IssuesListForRepoResponseData } from '@octokit/types';

import octokit from '@/utils/octokit';

import { IIssue } from '@/components/RepositoryPage/state/typings';
import { ArrayElement } from '@/typings/utils';

const mapDataItemToIssue = (
  item: ArrayElement<IssuesListForRepoResponseData>
): IIssue => ({
  id: item.node_id,
  title: item.title,
  link: item.html_url,
  state: item.state as 'open' | 'closed',
});

const fetchRepoIssues = async (
  owner: string,
  repo: string
): Promise<Array<IIssue>> => {
  const { data } = await octokit.issues.listForRepo({ owner, repo });
  return data.map(mapDataItemToIssue);
};

export default fetchRepoIssues;
