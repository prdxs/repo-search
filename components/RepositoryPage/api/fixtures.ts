import { IIssue } from '@/components/RepositoryPage/state/typings';

export const issues: Array<IIssue> = [
  {
    id: '0',
    title: "Bug: what's sky-lighter.appspot.com?",
    link: 'https://github.com/facebook/react/issues/19245',
    state: 'open',
  },
  {
    id: '1',
    title:
      "Bug: Cannot read property 'references' of undefined in eslint-plugin-react-hooks v4.0.5",
    link: 'https://github.com/facebook/react/issues/19243',
    state: 'closed',
  },
];
