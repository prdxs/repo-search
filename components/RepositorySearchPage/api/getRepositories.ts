import { SearchReposResponseData } from '@octokit/types';
import moment from 'moment';

import { ArrayElement } from '@/typings/utils';
import octokit from '@/utils/octokit';

import { IRepository } from '@/components/RepositorySearchPage/state/typings';

const mapDataItemsToRepos = (
  item: ArrayElement<SearchReposResponseData['items']>
): IRepository => ({
  id: item.node_id,
  name: item.name,
  stars: item.stargazers_count,
  link: item.html_url,
  forks: item.forks_count,
  issues: item.open_issues_count,
  lastUpdated: moment(item.updated_at).format('Do MMM YYYY HH:mm:ss'),
});

const getRepositories = async (q: string): Promise<Array<IRepository>> => {
  const { data: dataByUser } = await octokit.search.repos({ q: `user:${q}` });
  const { data: dataByKeyword } = await octokit.search.repos({ q });

  const byUserIds = dataByUser.items.map((item) => item.node_id);
  const reposByUser = dataByUser.items.map(mapDataItemsToRepos);
  const reposByKeyword = dataByKeyword.items
    .filter((item) => byUserIds.includes(item.node_id))
    .map(mapDataItemsToRepos);

  return [...reposByUser, ...reposByKeyword];
};

export default getRepositories;
