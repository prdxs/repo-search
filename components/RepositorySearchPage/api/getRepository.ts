import { ReposGetResponseData } from '@octokit/types';
import moment from 'moment';

import octokit from '@/utils/octokit';

import { IRepository } from '@/components/RepositorySearchPage/state/typings';

const mapDataItemToRepo = (item: ReposGetResponseData): IRepository => ({
  id: item.node_id,
  name: item.name,
  stars: item.stargazers_count,
  link: item.html_url,
  forks: item.forks_count,
  issues: item.open_issues_count,
  lastUpdated: moment(item.updated_at).format('Do MMM YYYY HH:mm:ss'),
});

const getRepository = async (
  owner: string,
  repo: string
): Promise<IRepository> => {
  const { data } = await octokit.repos.get({ owner, repo });
  return mapDataItemToRepo(data);
};

export default getRepository;
