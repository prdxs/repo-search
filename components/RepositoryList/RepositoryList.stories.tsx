import React from 'react';

import StorybookWrapper from '@/components/StorybookWrapper';
import RepositoryList from './RepositoryList.component';
import { repositories } from './api/fixtures';

export default {
  title: 'RepositoryList',
  component: RepositoryList,
};

export const Default: React.FC = () => {
  return (
    <StorybookWrapper>
      <RepositoryList repositories={repositories} />
    </StorybookWrapper>
  );
};
