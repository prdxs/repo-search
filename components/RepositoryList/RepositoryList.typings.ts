import { IComponentProps } from '@/typings/common';
import { IRepository } from './api/typings';

export interface IRepositoryListProps extends IComponentProps {
  repositories: Array<IRepository>;
}
