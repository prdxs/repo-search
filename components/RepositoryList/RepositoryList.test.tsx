import React from 'react';

import { render } from '@/test/utils';
import RepositoryList from './RepositoryList.component';
import { repositories } from './api/fixtures';

describe('RepositoryList', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(
      <RepositoryList repositories={repositories} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render as many repositories as provided', () => {
    const { getAllByTestId } = render(
      <RepositoryList repositories={repositories} />
    );

    const repositoryEls = getAllByTestId('RepositoryCard-root');

    expect(repositoryEls).toHaveLength(2);
  });
});
