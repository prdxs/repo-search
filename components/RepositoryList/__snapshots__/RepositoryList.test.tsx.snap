// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepositoryList should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
    style="padding: 2rem;"
  >
    <div
      class="MuiGrid-root RepositoryList RepositoryListcomponent__StyledRepositoryList-sc-12gcib3-0 kTpDEU MuiGrid-container MuiGrid-spacing-xs-4"
    >
      <div
        class="MuiGrid-root MuiGrid-item"
      >
        <div
          class="MuiPaper-root MuiCard-root RepositoryCard-root RepositoryCardcomponent__StyledRepositoryCard-sc-1c2lx1b-0 ljLEiR MuiPaper-elevation1 MuiPaper-rounded"
          data-testid="RepositoryCard-root"
        >
          <div
            class="MuiCardHeader-root"
          >
            <div
              class="MuiCardHeader-content"
            >
              <span
                class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
              >
                <div
                  class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1"
                >
                  <div
                    class="MuiGrid-root MuiGrid-item"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M18 2H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zM6 4h5v8l-2.5-1.5L6 12V4z"
                      />
                    </svg>
                  </div>
                  <div
                    class="MuiGrid-root MuiGrid-item"
                  >
                    <h6
                      class="MuiTypography-root MuiTypography-h6"
                    >
                      <a
                        class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorSecondary"
                        href="https://github.com/facebook/react"
                        target="_blank"
                      >
                        facebook/react
                      </a>
                    </h6>
                  </div>
                </div>
              </span>
            </div>
          </div>
          <div
            class="MuiCardContent-root"
          >
            <div
              class="MuiGrid-root MuiGrid-container"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
              >
                <div
                  class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
                >
                  <div
                    class="MuiGrid-root MuiGrid-item"
                  >
                    <span
                      class="MuiBadge-root"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root RepositoryCard-star MuiSvgIcon-fontSizeLarge"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                        />
                      </svg>
                      <span
                        class="MuiBadge-badge MuiBadge-anchorOriginBottomRightRectangle MuiBadge-colorPrimary"
                      >
                        151026
                      </span>
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
              >
                <div
                  class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
                >
                  <div
                    class="MuiGrid-root MuiGrid-item"
                  >
                    <span
                      class="MuiBadge-root"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M20 8h-2.81c-.45-.78-1.07-1.45-1.82-1.96L17 4.41 15.59 3l-2.17 2.17C12.96 5.06 12.49 5 12 5c-.49 0-.96.06-1.41.17L8.41 3 7 4.41l1.62 1.63C7.88 6.55 7.26 7.22 6.81 8H4v2h2.09c-.05.33-.09.66-.09 1v1H4v2h2v1c0 .34.04.67.09 1H4v2h2.81c1.04 1.79 2.97 3 5.19 3s4.15-1.21 5.19-3H20v-2h-2.09c.05-.33.09-.66.09-1v-1h2v-2h-2v-1c0-.34-.04-.67-.09-1H20V8zm-6 8h-4v-2h4v2zm0-4h-4v-2h4v2z"
                        />
                      </svg>
                      <span
                        class="MuiBadge-badge MuiBadge-anchorOriginBottomRightRectangle MuiBadge-colorSecondary"
                      >
                        498
                      </span>
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
              >
                <div
                  class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
                >
                  <div
                    class="MuiGrid-root MuiGrid-item"
                  >
                    <span
                      class="MuiBadge-root"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M14 4l2.29 2.29-2.88 2.88 1.42 1.42 2.88-2.88L20 10V4zm-4 0H4v6l2.29-2.29 4.71 4.7V20h2v-8.41l-5.29-5.3z"
                        />
                      </svg>
                      <span
                        class="MuiBadge-badge MuiBadge-anchorOriginBottomRightRectangle MuiBadge-colorSecondary"
                      >
                        29654
                      </span>
                    </span>
                  </div>
                </div>
              </div>
            </div>
            <span
              class="MuiTypography-root RepositoryCard-lastUpdated MuiTypography-caption"
            >
              1st Jul 2020 07:20:59
            </span>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item"
      >
        <div
          class="MuiPaper-root MuiCard-root RepositoryCard-root RepositoryCardcomponent__StyledRepositoryCard-sc-1c2lx1b-0 ljLEiR MuiPaper-elevation1 MuiPaper-rounded"
          data-testid="RepositoryCard-root"
        >
          <div
            class="MuiCardHeader-root"
          >
            <div
              class="MuiCardHeader-content"
            >
              <span
                class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
              >
                <div
                  class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1"
                >
                  <div
                    class="MuiGrid-root MuiGrid-item"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M18 2H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zM6 4h5v8l-2.5-1.5L6 12V4z"
                      />
                    </svg>
                  </div>
                  <div
                    class="MuiGrid-root MuiGrid-item"
                  >
                    <h6
                      class="MuiTypography-root MuiTypography-h6"
                    >
                      <a
                        class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorSecondary"
                        href="https://github.com/prdxs/repo-search"
                        target="_blank"
                      >
                        prdxs/repo-search
                      </a>
                    </h6>
                  </div>
                </div>
              </span>
            </div>
          </div>
          <div
            class="MuiCardContent-root"
          >
            <div
              class="MuiGrid-root MuiGrid-container"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
              >
                <div
                  class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
                >
                  <div
                    class="MuiGrid-root MuiGrid-item"
                  >
                    <span
                      class="MuiBadge-root"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root RepositoryCard-star MuiSvgIcon-fontSizeLarge"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                        />
                      </svg>
                      <span
                        class="MuiBadge-badge MuiBadge-anchorOriginBottomRightRectangle MuiBadge-colorPrimary"
                      >
                        0
                      </span>
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
              >
                <div
                  class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
                >
                  <div
                    class="MuiGrid-root MuiGrid-item"
                  >
                    <span
                      class="MuiBadge-root"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M20 8h-2.81c-.45-.78-1.07-1.45-1.82-1.96L17 4.41 15.59 3l-2.17 2.17C12.96 5.06 12.49 5 12 5c-.49 0-.96.06-1.41.17L8.41 3 7 4.41l1.62 1.63C7.88 6.55 7.26 7.22 6.81 8H4v2h2.09c-.05.33-.09.66-.09 1v1H4v2h2v1c0 .34.04.67.09 1H4v2h2.81c1.04 1.79 2.97 3 5.19 3s4.15-1.21 5.19-3H20v-2h-2.09c.05-.33.09-.66.09-1v-1h2v-2h-2v-1c0-.34-.04-.67-.09-1H20V8zm-6 8h-4v-2h4v2zm0-4h-4v-2h4v2z"
                        />
                      </svg>
                      <span
                        class="MuiBadge-badge MuiBadge-anchorOriginBottomRightRectangle MuiBadge-colorSecondary"
                      >
                        0
                      </span>
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
              >
                <div
                  class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
                >
                  <div
                    class="MuiGrid-root MuiGrid-item"
                  >
                    <span
                      class="MuiBadge-root"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M14 4l2.29 2.29-2.88 2.88 1.42 1.42 2.88-2.88L20 10V4zm-4 0H4v6l2.29-2.29 4.71 4.7V20h2v-8.41l-5.29-5.3z"
                        />
                      </svg>
                      <span
                        class="MuiBadge-badge MuiBadge-anchorOriginBottomRightRectangle MuiBadge-colorSecondary"
                      >
                        0
                      </span>
                    </span>
                  </div>
                </div>
              </div>
            </div>
            <span
              class="MuiTypography-root RepositoryCard-lastUpdated MuiTypography-caption"
            >
              1st Jul 2020 14:20:17
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
